lld/MachO/BPSectionOrderer.h

```diff
@@ -14,13 +14,145 @@
 #ifndef LLD_MACHO_BPSECTION_ORDERER_H
 #define LLD_MACHO_BPSECTION_ORDERER_H
 
+#include "InputSection.h"
+#include "Relocations.h"
+#include "Symbols.h"
+#include "lld/Common/BPSectionOrdererBase.h"
 #include "llvm/ADT/DenseMap.h"
 #include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/TinyPtrVector.h"
 
 namespace lld::macho {
 
 class InputSection;
 
+class BPSymbolMacho : public BPSymbol {
+  const Symbol *sym;
+
+public:
+  explicit BPSymbolMacho(const Symbol *s) : sym(s) {}
+
+  llvm::StringRef getName() const override { return sym->getName(); }
+
+  const Defined *asDefined() const {
+    return llvm::dyn_cast_or_null<Defined>(sym);
+  }
+
+  std::optional<uint64_t> getValue() const override {
+    if (auto *d = asDefined())
+      return d->value;
+    return {};
+  }
+
+  std::optional<uint64_t> getSize() const override {
+    if (auto *d = asDefined())
+      return d->size;
+    return {};
+  }
+
+  const Symbol *getSymbol() const { return sym; }
+};
+
+class BPSectionMacho : public BPSectionBase {
+  const InputSection *isec;
+  uint64_t sectionIdx;
+
+public:
+  explicit BPSectionMacho(const InputSection *sec, uint64_t sectionIdx)
+      : isec(sec), sectionIdx(sectionIdx) {}
+
+  const InputSection *getSection() const { return isec; }
+
+  uint64_t getSize() const override { return isec->getSize(); }
+
+  uint64_t getSectionIdx() const { return sectionIdx; }
+
+  bool isCodeSection() const override { return macho::isCodeSection(isec); }
+
+  bool hasValidData() const override {
+    return isec && !isec->data.empty() && isec->data.data();
+  }
+
+  SmallVector<std::unique_ptr<BPSymbol>> getSymbols() const override {
+    SmallVector<std::unique_ptr<BPSymbol>> symbols;
+    for (auto *sym : isec->symbols)
+      if (auto *d = llvm::dyn_cast_or_null<Defined>(sym))
+        symbols.emplace_back(std::make_unique<BPSymbolMacho>(d));
+    return symbols;
+  }
+
+  // Linkage names can be prefixed with "_" or "l_" on Mach-O. See
+  // Mangler::getNameWithPrefix() for details.
+  std::optional<StringRef>
+  getResolvedLinkageName(llvm::StringRef name) const override {
+    if (name.consume_front("_") || name.consume_front("l_"))
+      return name;
+    return {};
+  }
+
+  void getSectionHashes(llvm::SmallVectorImpl<uint64_t> &hashes,
+                        const llvm::DenseMap<const BPSectionBase *, uint64_t>
+                            &sectionToIdx) const override {
+    constexpr unsigned windowSize = 4;
+
+    // Create a map from BPSectionBase* to InputSection* for lookup
+    llvm::DenseMap<const InputSection *, uint64_t> isecToIdx;
+    for (const auto &[section, idx] : sectionToIdx)
+      if (auto *machoSection = llvm::dyn_cast<BPSectionMacho>(section))
+        isecToIdx[machoSection->getSection()] = idx;
```

Comment:
Sorry, this is still not efficient because we call `getSectionHashes` for each section, making this `O(n^2)`.

I guess we only need `sectionToIdx` for Mach-O because ELF doesn't take relocations into account. Can we somehow make `DenseMap<const InputSection *, uint64_t> sectionToIdx` specific to Mach-O? Or should we make the key type `void *` and cast to the Mach-O or ELF version of `InputSection *` when we need it?

By the way, have you tried to hash relocations on ELF? My guess is that you could get a pretty nice uncompressed and compressed size win.

Reply:
I'm trying to use void * for this. Based on our MachO hash relocation results, I think it will perform quite well on ELF too. I'll spend some time refining it.

Reply:
@ellishg Hi ellis, considering the complexity of implementing ELF relocations, I think the hash relocations for ELF may take some time to complete. How about using the simple sectionHash as a base for now and adding relocation optimizations in the future? What do you think?

Reply:
I'm fine with that for ELF, but we don't want to regress anything for Mach-O

Reply:
Got it. I'll ensure there are no regressions for Mach-O while doing ELF relocations hashes.